src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -641,11 +641,11 @@ public final IConditionExpr getConditionExpressionForTrueAction(FormInstance mai
      * API for retrieving the list of conditions, for use when
      * serializing the form definition (e.g., into .cache file).
      */
-    public List<Condition> getConditions() {
-        List<Condition> conditions = new ArrayList<>();
+    public List<Triggerable> getConditions() {
```

Comment:
This doesn't seem right. In this case, we really do want only conditions and I think the type should indicate that.

Reply:
I think it's fine. This method is part of the external serialization code, which is kind of an internal detail that we have tight control over. We could even inline that method and we wouldn't have to worry about it.

Anyway, after studying the code, I see there are some confusing concepts related to `Condition` that we might want to come back to and review. The best example of confusing code is when we ask for a condition when evaluating the relevance of an element and we implicitly assume that we're asking for the _relevance_ condition when the DAG could contain any number of different conditions too. Check `FormDef.isRepeatRelevant` and `TriggerableDag.getTriggerableForRepeatGroup`.

At this moment, here's my rationale:
- Our domain talks about "Relevant", "Required", "Constraint", "Read Only", and "Calculate", not about "Condition".

  Previously, these domain concepts were gathered by variable and member names, which is fine for now, I think. I don't think we're losing domain detail by using `Triggerable` here.

  Looking forward, I think we could want to create classes for all the domain concepts for type checking only. Maybe for object creation too, since we already have established that the target reference set for a relevance condition, when applied to a group, should include all descendant refs (kinda different behavior we could want to move into the type).

- Regarding the high-level (abstract) behavior of `Condition`, and `Recalculate`, I don't see any difference: they are triggered, they are eval'ed, and side-effects of their evaluation are applied to some targets.

- We control the creation of triggerables in the internals of the DAG, which is less risky than having this as part of the public API, where having static type checking would be super useful.

Reply:
So, in the end, I refactored the serialization code to make those method private and now I'm confortable returning `Condition` and `Recalculate`. Let me know what you think ;)

Reply:
Looks great to me, thanks!